File: src/c/ch01/1.3/l2/miller_rabin.c
#include "miller_rabin.h"

/* Safe modular multiply using 128-bit intermediate (GCC/Clang on mainstream 64-bit). */
static inline uint64_t mulmod_u64(uint64_t a, uint64_t b, uint64_t m) {
    __uint128_t t = ( (__uint128_t)a * (__uint128_t)b ) % m;
    return (uint64_t)t;
}

/* Fast modular exponentiation. */
static uint64_t powmod_u64(uint64_t base, uint64_t exp, uint64_t mod) {
    if (mod == 1) return 0;
    uint64_t res = 1 % mod;
    uint64_t b = base % mod;
    while (exp > 0) {
        if (exp & 1) res = mulmod_u64(res, b, mod);
        b = mulmod_u64(b, b, mod);
        exp >>= 1;
    }
    return res;
}

static inline void decompose(uint64_t n_minus_1, uint64_t *d, uint32_t *s) {
    uint64_t dd = n_minus_1;
    uint32_t ss = 0;
    while ((dd & 1ULL) == 0) { dd >>= 1; ++ss; }
    *d = dd; *s = ss;
}

static bool try_composite(uint64_t a, uint64_t d, uint64_t n, uint32_t s) {
    uint64_t x = powmod_u64(a, d, n);
    if (x == 1 || x == n - 1) return false;
    for (uint32_t i = 1; i < s; ++i) {
        x = mulmod_u64(x, x, n);
        if (x == n - 1) return false;
    }
    return true; /* definitely composite for this base */
}

bool is_probable_prime_u64(uint64_t n) {
    if (n < 2) return false;
    /* small primes & quick divisibility */
    const uint32_t small_primes[] = {2,3,5,7,11,13,17,19,23,29};
    for (unsigned i = 0; i < sizeof(small_primes)/sizeof(small_primes[0]); ++i) {
        uint32_t p = small_primes[i];
        if (n == p) return true;
        if (n % p == 0) return false;
    }

    uint64_t d; uint32_t s;
    decompose(n - 1, &d, &s);

    /* Deterministic base set for 64-bit range. */
    const uint64_t bases[] = {2ULL, 3ULL, 5ULL, 7ULL, 11ULL, 13ULL, 17ULL};
    for (unsigned i = 0; i < sizeof(bases)/sizeof(bases[0]); ++i) {
        uint64_t a = bases[i] % n;
        if (a == 0) continue;
        if (try_composite(a, d, n, s)) return false;
    }
    return true;
}

File: archived/v1/src/go/ch01/1.3/l2/miller_rabin.go
package main

import (
	"math/big"
)

// powModBig computes (a^e mod n) with big.Int (handles overflow safely).
func powModBig(a, e, n *big.Int) *big.Int {
	var res big.Int
	res.SetUint64(1)
	var base big.Int
	base.Mod(a, n)
	var exp big.Int
	exp.Set(e)

	for exp.Sign() > 0 {
		if exp.Bit(0) == 1 {
			res.Mul(&res, &base).Mod(&res, n)
		}
		base.Mul(&base, &base).Mod(&base, n)
		exp.Rsh(&exp, 1)
	}
	return &res
}

func decompose(nMinus1 uint64) (d uint64, s uint32) {
	d = nMinus1
	s = 0
	for (d & 1) == 0 {
		d >>= 1
		s++
	}
	return
}

func tryComposite(base uint64, d uint64, n uint64, s uint32) bool {
	N := new(big.Int).SetUint64(n)
	A := new(big.Int).SetUint64(base % n)
	D := new(big.Int).SetUint64(d)

	x := powModBig(A, D, N)
	if x.Cmp(big.NewInt(1)) == 0 || x.Cmp(new(big.Int).Sub(N, big.NewInt(1))) == 0 {
		return false
	}
	for i := uint32(1); i < s; i++ {
		x.Mul(x, x).Mod(x, N)
		if x.Cmp(new(big.Int).Sub(N, big.NewInt(1))) == 0 {
			return false
		}
	}
	return true
}

// IsProbablePrimeU64: Miller–Rabin deterministic for 64-bit with fixed bases.
func IsProbablePrimeU64(n uint64) bool {
	if n < 2 {
		return false
	}
	small := []uint64{2, 3, 5, 7, 11, 13, 17, 19, 23, 29}
	for _, p := range small {
		if n == p {
			return true
		}
		if n%p == 0 {
			return false
		}
	}
	d, s := decompose(n - 1)
	bases := []uint64{2, 3, 5, 7, 11, 13, 17}
	for _, a := range bases {
		aa := a % n
		if aa == 0 {
			continue
		}
		if tryComposite(aa, d, n, s) {
			return false
		}
	}
	return true
}